internal/xds/translator/translator.go

```diff
@@ -113,29 +114,57 @@ func Translate(ir *ir.Xds) (*types.ResourceVersionTable, error) {
 		tCtx.AddXdsResource(resource.ClusterType, xdsCluster)
 
 		// Search for an existing listener, if it does not exist, create one.
-		xdsListener := findXdsListener(tCtx, tcpListener.Address, tcpListener.Port)
+		xdsListener := findXdsListener(tCtx, tcpListener.Address, tcpListener.Port, core.SocketAddress_TCP)
 		if xdsListener == nil {
-			xdsListener = buildXdsListener(tcpListener.Name, tcpListener.Address, tcpListener.Port)
+			xdsListener = buildXdsTCPListener(tcpListener.Name, tcpListener.Address, tcpListener.Port)
 			tCtx.AddXdsResource(resource.ListenerType, xdsListener)
 		}
-
+		// todo: huabing zhao we should check the conflicted listeners: multiple TCP listeners or TCP and HTTP listeners
+		// should not listen on the same port
 		if err := addXdsTCPFilterChain(xdsListener, tcpListener, xdsCluster.Name); err != nil {
 			return nil, err
 		}
 	}
+
+	for _, udpListener := range ir.UDP {
+		// 1:1 between IR UDPListener and xDS Cluster
+		xdsCluster, err := buildXdsCluster(udpListener.Name, udpListener.Destinations)
+		if err != nil {
+			return nil, multierror.Append(err, errors.New("error building xds cluster"))
+		}
+		tCtx.AddXdsResource(resource.ClusterType, xdsCluster)
+
+		// Search for an existing listener, if it does not exist, create one.
+		xdsListener := findXdsListener(tCtx, udpListener.Address, udpListener.Port, core.SocketAddress_UDP)
```

Comment:
why is the `find` needed here, is there any case where we want to support `udp` and another protocol on the same addr:port ?

Reply:
No, but we need to throw an error in case there are two listeners on the same UDP ports.

Reply:
i think this case should be handled in the Gateway API translator itself, which has enough info to detect this as well as signal back to the user using the `UDPRoute` `Status` sub resource, prefer if this code is simpler

Reply:
> u also enhance this PR with ac

sure

Reply:
> i think this case should be handled in the Gateway API translator itself, which has enough info to detect this as well as signal back to the user using the `UDPRoute` `Status` sub resource, prefer if this code is simpler

OK, we can remove the check here if Gateway API translator can detect it in the first place.